### PR TITLE
consolidate RouteRequestCallback and Router.Callback into a single RouterCallback with improved error accessibility

### PR DESCRIPTION
**Pull request by LukasPaczos**
### Description

Refs https://github.com/mapbox/mapbox-navigation-android/issues/4376. This contribution refactors our route requests callback, removes the duplication in the public API and exposes much more information about the reason for the failure, including the original request URL and HTTP codes and error bodies (when available).

In short, we're conslidating:
```kotlin
interface Router.Callback {
    fun onResponse(routes: List<DirectionsRoute>)
    fun onFailure(throwable: Throwable)
    fun onCanceled()
}
```
and
```kotlin
interface RoutesRequestCallback {
    fun onRoutesReady(routes: List<DirectionsRoute>)
    fun onRoutesRequestFailure(throwable: Throwable, routeOptions: RouteOptions)
    fun onRoutesRequestCanceled(routeOptions: RouteOptions)
}
```
that were used in different parts of the codebase (and ultimately did the same thing) into:
```kotlin
interface RouterCallback {
    fun onRoutesReady(routes: List<DirectionsRoute>)
    fun onFailure(reasons: List<RouterFailure>, routeOptions: RouteOptions)
    fun onCanceled(routeOptions: RouteOptions)
}
```

where `RouterFailure` provides a code and reason, also for nested failures.
```kotlin
data class RouterFailure(
    val url: URL,
    val message: String,
    val code: Int? = null,
    val throwable: Throwable? = null
)
```
For example, if both offboard and onboard routers fail, we now have a way to provide both reasons.

The next steps are to expose the [Directions API error codes](https://docs.mapbox.com/api/navigation/directions/#directions-api-errors) as string defs or enums to fully close https://github.com/mapbox/mapbox-navigation-android/issues/4376, and make the similar refresh interfaces internal and hidden since they are not used outside of the SDK scope.

### Changelog

```
<changelog>:warning: Removed `RouteRequestCallback` and `Router.Callback` in favor of a single `RouterCallback` with improved failure reason accessibility.</changelog>
```

### CODE REVIEW

File: libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/offboard/MapboxOffboardRouter.kt [hunk 1]
```diff
@@ -70,22 +69,48 @@ class MapboxOffboardRouter(
...
+                                        RouterFailure(url, "routes list is empty", code)
+                                    ),
+                                    routeOptions
+                                )
+                            }
                         }
-                        else -> callback.onFailure(NavigationException(response.message()))
+                        else -> callback.onFailure(
+                            listOf(
+                                RouterFailure(url, response.errorBody()?.string().toString(), code)
```
Comment (korshaknn) [inline libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/offboard/MapboxOffboardRouter.kt:91]:
```suggestion
                                RouterFailure(url, response.errorBody()?.string()), code)
```
can be simplified?

File: libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/offboard/MapboxOffboardRouter.kt [hunk 2]
```diff
@@ -70,22 +69,48 @@ class MapboxOffboardRouter(
...
                     if (call.isCanceled) {
-                        callback.onCanceled()
+                        callback.onCanceled(routeOptions)
                     } else {
-                        callback.onFailure(t)
+                        callback.onFailure(
+                            listOf(
+                                RouterFailure(
+                                    url = call.request().url.toUrl(),
+                                    message = t.message ?: "unknown",
```
Comment (korshaknn) [inline libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/offboard/MapboxOffboardRouter.kt:107]: move  "unknown" to constants

File: libnavigation-router/src/test/java/com/mapbox/navigation/route/internal/offboard/MapboxOffboardRouterTest.kt [hunk 3]
```diff
@@ -163,64 +173,80 @@ class MapboxOffboardRouterTest : BaseTest() {
...
     @Test
     fun onFailure_errorIsProvided() {
-        val routerCallback = mockk<Router.Callback>(relaxed = true)
-        val throwable = mockk<Throwable>()
+        val routerCallback = mockk<RouterCallback>(relaxed = true)
+        val url = HttpUrl.Builder()
+            .scheme("https")
+            .host("api.mapbox.com")
+            .addPathSegment("test")
+            .build()
```
Comment (korshaknn) [inline libnavigation-router/src/test/java/com/mapbox/navigation/route/internal/offboard/MapboxOffboardRouterTest.kt:227]: there is a global 
```
private val url = HttpUrl.Builder()
        .scheme("https")
        .host("api.mapbox.com")
        .addPathSegment("test")
        .build()
```
can the local be removed?

**Review by korshaknn — COMMENTED** [review]
great work! added some comments

File: examples/src/main/java/com/mapbox/navigation/examples/core/camera/MapboxCameraAnimationsActivity.kt [hunk 4]
```diff
@@ -556,24 +557,24 @@ class MapboxCameraAnimationsActivity :
             .applyDefaultNavigationOptions()
             .applyLanguageAndVoiceUnitOptions(this)
             .accessToken(getMapboxAccessTokenFromResources())
-            .coordinates(listOf(origin, destination))
+            .coordinates(listOf(origin, Point.fromLngLat(-20.171554878847626, 58.058206980462955)))
```
[the remark this answers is not in the document]
Comment (LukasPaczos) [reply examples/src/main/java/com/mapbox/navigation/examples/core/camera/MapboxCameraAnimationsActivity.kt:560]: Good catch, I was testing something else :upside_down_face:

File: libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/offboard/MapboxOffboardRouter.kt (hunk 1 above)
Comment (LukasPaczos) [reply libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/offboard/MapboxOffboardRouter.kt:91]: `ResponseBody#string` is nullable, that's why we need the extension

**Comment by LukasPaczos** [comment]
Ready for another round.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt [hunk 5]
```diff
@@ -83,37 +84,34 @@ internal class MapboxDirectionsSession(
...
-                            """
-                                Route request - failure: $throwable
-                                For options: $routeOptions
-                            """.trimIndent()
+                override fun onFailure(reasons: List<RouterFailure>, routeOptions: RouteOptions) {
+                    reasons.forEach {
+                        LoggerProvider.logger.e(
+                            Tag(TAG),
+                            Message(it.toString())
                         )
```
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt:108]: Do we really want to log out now that the `reasons` are available? 🤔 

Same thing applies to `onCanceled`.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/MapboxRerouteController.kt [hunk 6]
```diff
@@ -129,23 +130,22 @@ internal class MapboxRerouteController(
                     }
                 }
 
-                override fun onRoutesRequestFailure(
-                    throwable: Throwable,
+                override fun onFailure(
+                    reasons: List<RouterFailure>,
                     routeOptions: RouteOptions
                 ) {
                     logger.e(
```
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/MapboxRerouteController.kt:137]: In any case, we're already logging in `MapboxDirectionsSession`, do we want to log again here?

Same for `onCanceled` 👇

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/RerouteController.kt [hunk 7]
```diff
@@ -98,9 +99,14 @@ sealed class RerouteState {
      * You can [MapboxNavigation.requestRoutes] or [MapboxNavigation.setRoutes] with [MapboxRouteOptionsUpdater] to retry the request.
      *
      * @param message describes error
-     * @param throwable is optional
+     * @param throwable optional throwable
+     * @param reasons optional reasons for the failure
      */
-    data class Failed(val message: String, val throwable: Throwable? = null) : RerouteState()
+    data class Failed(
```
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/RerouteController.kt:105]:
```suggestion
    data class Failed @JvmOverloads constructor(
```

File: libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/hybrid/HybridRouterHandler.kt [hunk 8]
```diff
@@ -44,79 +45,83 @@ internal sealed class HybridRouterHandler(
...
+            private val clientCallback: RouterCallback
+        ) : RouterCallback {
+            override fun onRoutesReady(routes: List<DirectionsRoute>) {
+                clientCallback.onRoutesReady(routes)
             }
 
-            override fun onFailure(throwable: Throwable) {
-                logger.w(
+            override fun onFailure(reasons: List<RouterFailure>, routeOptions: RouteOptions) {
+                LoggerProvider.logger.w(
```
Comment (Guardiola31337) [inline libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/hybrid/HybridRouterHandler.kt:58]: As mentioned above, we're also logging here. Do we need all this logging "duplication"?

File: libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/offboard/MapboxOffboardRouter.kt [hunk 9]
```diff
@@ -70,22 +71,48 @@ class MapboxOffboardRouter(
...
+                            if (!routes.isNullOrEmpty()) {
+                                callback.onRoutesReady(routes)
+                            } else {
+                                callback.onFailure(
+                                    listOf(
+                                        RouterFailure(url, ROUTES_LIST_EMPTY, code)
+                                    ),
+                                    routeOptions
+                                )
+                            }
```
Comment (Guardiola31337) [inline libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/offboard/MapboxOffboardRouter.kt:89]: Didn't we agree that this is not an option? cc @danpat @mskurydin @vanmaxim

File: libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/offboard/MapboxOffboardRouter.kt [hunk 10]
```diff
@@ -70,22 +71,48 @@ class MapboxOffboardRouter(
                     response: Response<DirectionsResponse>
                 ) {
                     directionRequests.remove(requestId)
+                    val url = call.request().url.toUrl()
```
Comment (Guardiola31337) [inline libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/offboard/MapboxOffboardRouter.kt:74]: For "security" reasons, do we want to filter out the `access_token` from the URL so it's not leaked out in the Logcat?

File: libnavigation-router/src/main/java/com/mapbox/navigation/route/offboard/router/ElapsedTime.kt [hunk 11]
```diff
@@ -17,7 +16,7 @@ internal class ElapsedTime(
 
     fun end() {
         if (start == null) {
-            throw NavigationException("Must call start() before calling end()")
+            throw RuntimeException("Must call start() before calling end()")
```
Comment (Guardiola31337) [inline libnavigation-router/src/main/java/com/mapbox/navigation/route/offboard/router/ElapsedTime.kt:19]:
```suggestion
            throw IllegalStateException("Must call start() before calling end()")
```

**Review by Guardiola31337 — COMMENTED** [review]
Great work @LukasPaczos 

Left some comments to discuss before merging here.

**Comment by Guardiola31337** [comment]
@LukasPaczos 
> make the similar refresh interfaces internal and hidden since they are not used outside of the SDK scope

You mean refactor _Refresh_ APIs similarly to `Router` ones and provide same consistent messages and failures, right?

**Comment by LukasPaczos** [comment]
I meant moving the refresh interfaces from `base` to `core` and making them internal (as much as possible), or just moving them to `internal` package and outside of the public API. Those interfaces are only used internally. We can refactor them similarly though so that we can log out the refresh failure better.

File: libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/offboard/MapboxOffboardRouter.kt (hunk 10 above)
Comment (LukasPaczos) [reply libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/offboard/MapboxOffboardRouter.kt:74]: :+1:

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt (hunk 5 above)
Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt:108]: Probably it's not necessary anymore :+1:

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/route/Router.kt [hunk 12]
```diff
@@ -59,29 +60,20 @@ interface Router {
...
+/**
+ * Describes a reason for a route request failure.
+ *
+ * @param url original request URL
+ * @param message message attached to the error code
+ * @param code if present, can be either be the HTTP code for offboard requests
+ * or an internal error code for onboard requests
+ * @param throwable provided if an unexpected exception occurred when creating the request or processing the response
+ */
+data class RouterFailure(
```
[the remark this answers is not in the document]
Comment (LukasPaczos) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/route/Router.kt:74]: If we won't log this out by default, I think adding a separate field makes sense :+1:

**Comment by Guardiola31337** [comment]
Yeah that makes total sense 👍 

> We can refactor them similarly though so that we can log out the refresh failure better.

Yeah, for consistency would like to do this too (not blocking this PR though) 👍

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt (hunk 5 above)
Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt:108]: Alternatively, we can decrease the log level to keep this around by default but not pollute the top log levels as much.

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/route/Router.kt (hunk 12 above)
[the remark this answers is not in the document]
Comment (LukasPaczos) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/route/Router.kt:74]: Let me follow up on online/offline changes in a separate PR to not overload this one. We can then add the flag to the failure reason as well.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/MapboxRerouteController.kt (hunk 6 above)
Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/MapboxRerouteController.kt:137]: I think this one should stay. Developers can't hook into a failure of our default reroute controller, so we need to have a way to communicate the errors.

Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/MapboxRerouteController.kt:137]: Ah this is not true, developers can use `mapboxNavigation.getRerouteController()?.registerRerouteStateObserver` :+1:

File: libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/hybrid/HybridRouterHandler.kt (hunk 8 above)
Comment (LukasPaczos) [reply libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/hybrid/HybridRouterHandler.kt:58]: I think this one needs to stay - if the primary route fails but the fallback succeeds, developers would never know that primary failed. This log is the only way to get this communicated.

File: libnavigation-router/src/main/java/com/mapbox/navigation/route/offboard/router/ElapsedTime.kt (hunk 11 above)
Comment (LukasPaczos) [reply libnavigation-router/src/main/java/com/mapbox/navigation/route/offboard/router/ElapsedTime.kt:19]: Removed altogether, as we discussed last time it's not used.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt (hunk 5 above)
Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt]: Opted to remove.

**Comment by LukasPaczos** [comment]
Ready for another round.

**Review by korshaknn — APPROVED** [review]
LGTM

File: libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/util/HttpUrlEx.kt [hunk 13]
```diff
@@ -0,0 +1,26 @@
+package com.mapbox.navigation.route.internal.util
+
+import okhttp3.HttpUrl
+import okhttp3.HttpUrl.Companion.toHttpUrl
+
+internal const val ACCESS_TOKEN_QUERY_PARAM = "access_token"
```
Comment (RingerJK) [inline libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/util/HttpUrlEx.kt:6]: add pls `todo move to mapbox-java`

File: libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/util/HttpUrlEx.kt [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+internal const val REDACTED = "redacted"
+
+internal fun String.redactQueryParam(queryParameterName: String): String {
```
Comment (RingerJK) [inline libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/util/HttpUrlEx.kt:9]: could you add doc that String should be URL

**Review by Guardiola31337 — APPROVED** [review]
Thanks for addressing the feedback :shipit:

File: libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/util/HttpUrlEx.kt (hunk 13 above)
Comment (LukasPaczos) [reply libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/util/HttpUrlEx.kt:6]: I'll follow up separately to not block the PR.

File: libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/util/HttpUrlEx.kt (hunk 14 above)
Comment (LukasPaczos) [reply libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/util/HttpUrlEx.kt:9]: I'll follow up separately to not block the PR.

File: libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/offboard/MapboxOffboardRouter.kt (hunk 9 above)
Comment (danpat) [reply libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/offboard/MapboxOffboardRouter.kt]: It should never occur - if it does, it's a server-side bug.

**Pull request merged**